File: lib-segment/src/main/java/websiteschema/mpsegment/core/SimpleConceptRecognizer.java
package websiteschema.mpsegment.core;

import websiteschema.mpsegment.concept.Concept;
import websiteschema.mpsegment.dict.IWord;
import websiteschema.mpsegment.dict.POSUtil;

public class SimpleConceptRecognizer implements IConceptRecognizer {
    private IWord[] words;
    private int[] posArray;

    @Override
    public void setWordArray(IWord[] wordArray) {
        this.words = wordArray;
    }

    @Override
    public void setPosArray(int[] posArray) {
        this.posArray = posArray;
    }

    @Override
    public String[] getConcepts() {
        assert (words != null && posArray != null && words.length == posArray.length);
        final int length = words.length;
        String[] concepts = new String[length];
        for(int i = 0; i < length; i++) {
            concepts[i] = getConcept(words[i], posArray[i]);
        }
        return concepts;
    }

    private String getConcept(IWord word, int pos) {
        Concept[] concepts = word.getConcepts();

        if(pos == POSUtil.POS_NS) {
            return "n-location";
        }

        if(pos == POSUtil.POS_NR) {
            return "n-name";
        }

        if (null != concepts) {
            String primaryPOS = POSUtil.getPOSString(pos).substring(0, 1).toLowerCase();
            for (int i = 0; i < concepts.length; i++) {
                if (concepts[i].getName().startsWith(primaryPOS)) {
                    return concepts[i].getName();
                }
            }
        }
        return Concept.UNKNOWN.getName();
    }

    @Override
    public void reset() {
        words = null;
        posArray = null;
    }


}

File: lib-segment/src/main/scala/websiteschema/mpsegment/core/SimpleConceptRecognizer.scala
package websiteschema.mpsegment.core

import websiteschema.mpsegment.concept.Concept
import websiteschema.mpsegment.dict.IWord
import websiteschema.mpsegment.dict.POSUtil

class SimpleConceptRecognizer extends IConceptRecognizer {
    private var words : Array[IWord] = null
    private var posArray : Array[Int] = null

    override def setWordArray(wordArray: Array[IWord]) {
        this.words = wordArray
    }

    override def setPosArray(posArray: Array[Int]) {
        this.posArray = posArray
    }

    override def getConcepts() : Array[String] = {
        assert (words != null && posArray != null && words.length == posArray.length)
        val length = words.length
        val concepts = new Array[String](length)
        for(i <- 0 until length) {
            concepts(i) = getConcept(words(i), posArray(i))
        }
        return concepts
    }

    private def getConcept(word: IWord, pos: Int) : String = {
        val concepts = word.getConcepts()

        if(pos == POSUtil.POS_NS) {
            return "n-location"
        }

        if(pos == POSUtil.POS_NR) {
            return "n-name"
        }

        if (null != concepts) {
            val primaryPOS = POSUtil.getPOSString(pos).substring(0, 1).toLowerCase()
            for (i <- 0 until concepts.length) {
                if (concepts(i).getName().startsWith(primaryPOS)) {
                    return concepts(i).getName()
                }
            }
        }
        return Concept.UNKNOWN.getName()
    }

    override def reset() {
        words = null
        posArray = null
    }


}
